Add test to assert that an async write failure fails the next write or stream-close.

saxenapranav opened the pull request:
Following methods: `write` ,  `close`, `hsync`, `hflush` method have reference to `maybeThrowLastError()` which is:
```
 private void maybeThrowLastError() throws IOException {
    if (lastError != null) {
      throw lastError;
    }
  }
```

how `lastError` is set?
1. Let there be 3 writes. First write fails and second write comes after the first failure: following code will set lastError which would fail next above-mentioned methods.
2. Lets say due to parallelism, no write started after any other write. It shall be caught in   `close`, `hsync`, `hflush`, `flush`. For `close` / `hsync` / `hflush`: it calls `flushInternal`:  it calls `flushWrittenBytesToService(isClose);` (https://github.com/ABFSDriver/AbfsHadoop/blob/450d524571e768b6a1171bd243a95bc7ef18bc68/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java#L696), which calls `waitForAppendsToComplete` (https://github.com/ABFSDriver/AbfsHadoop/blob/450d524571e768b6a1171bd243a95bc7ef18bc68/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java#L781): where we wait for all the writeOperations to complete, if failed, we throw the exception: https://github.com/ABFSDriver/AbfsHadoop/blob/450d524571e768b6a1171bd243a95bc7ef18bc68/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java#L759-L778. <br> In case of `flush`: ` flushInternalAsync();`: `waitForAppendsToComplete();` is there (https://github.com/ABFSDriver/AbfsHadoop/blob/450d524571e768b6a1171bd243a95bc7ef18bc68/hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java#L714)

What are in the tests?
1. Write which fails with 503, assert that next close() will raise the exception.
2. write which fails with 503, wait that the 503 has been received to outputStream, assert that next write() will raise the exception.
3. make two writes which fails with 503, but the second write is executed before first write fails. Assert that close will throw the exception.

saxenapranav commented:
------------------------
:::: AGGREGATED TEST RESULT ::::

HNS-OAuth
========================
[INFO] Results:
[INFO]
[WARNING] Tests run: 111, Failures: 0, Errors: 0, Skipped: 4
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   ITestAzureBlobFileSystemLease.testAcquireRetry:358 » TestTimedOut test timed o...
[INFO]
[ERROR] Tests run: 747, Failures: 0, Errors: 1, Skipped: 157
[INFO] Results:
[INFO]
[WARNING] Tests run: 273, Failures: 0, Errors: 0, Skipped: 41

HNS-SharedKey
========================
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   TestAbfsClientThrottlingAnalyzer.testManySuccessAndErrorsAndWaiting:181->fuzzyValidate:64 The actual value 9 is not within the expected range: [5.60, 8.40].
[INFO]
[ERROR] Tests run: 111, Failures: 1, Errors: 0, Skipped: 4
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   ITestAzureBlobFileSystemLease.testAcquireRetry:363 » TestTimedOut test timed o...
[INFO]
[ERROR] Tests run: 747, Failures: 0, Errors: 1, Skipped: 157
[INFO] Results:
[INFO]
[WARNING] Tests run: 273, Failures: 0, Errors: 0, Skipped: 41

NonHNS-SharedKey
========================
[INFO] Results:
[INFO]
[WARNING] Tests run: 111, Failures: 0, Errors: 0, Skipped: 4
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   ITestAzureBlobFileSystemLease.testAcquireRetry:358 » TestTimedOut test timed o...
[INFO]
[ERROR] Tests run: 747, Failures: 0, Errors: 1, Skipped: 275
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   ITestAbfsFileSystemContractDelete>AbstractContractDeleteTest.testDeleteNonEmptyDirNonRecursive:77->Assert.fail:89 non recursive delete should have raised an exception, but completed with exit code false
[ERROR] Errors:
[ERROR]   ITestAbfsFileSystemContractDistCp>AbstractContractDistCpTest.testTrackDeepDirectoryStructureToRemote:379->AbstractContractDistCpTest.distCpDeepDirectoryStructure:542->AbstractContractDistCpTest.lsR:472 » FileNotFound
[ERROR]   ITestAbfsFileSystemContractDistCp>AbstractContractDistCpTest.testUpdateDeepDirectoryStructureNoChange:262->AbstractContractDistCpTest.distCpDeepDirectoryStructure:543 » FileNotFound
[INFO]
[ERROR] Tests run: 273, Failures: 1, Errors: 2, Skipped: 45

NonHNS-OAuth
========================
[INFO] Results:
[INFO]
[WARNING] Tests run: 111, Failures: 0, Errors: 0, Skipped: 4
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   ITestAzureBlobFileSystemLease.testAcquireRetry:358 » TestTimedOut test timed o...
[INFO]
[ERROR] Tests run: 747, Failures: 0, Errors: 1, Skipped: 275
[INFO] Results:
[INFO]
[WARNING] Tests run: 273, Failures: 0, Errors: 0, Skipped: 45

AppendBlob-HNS-OAuth
========================
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]   TestAbfsClientThrottlingAnalyzer.testManySuccessAndErrorsAndWaiting:181->fuzzyValidate:64 The actual value 10 is not within the expected range: [5.60, 8.40].
[INFO]
[ERROR] Tests run: 111, Failures: 1, Errors: 0, Skipped: 4
[INFO] Results:
[INFO]
[ERROR] Errors:
[ERROR]   ITestAzureBlobFileSystemLease.testAcquireRetry:358 » TestTimedOut test timed o...
[INFO]
[ERROR] Tests run: 747, Failures: 0, Errors: 1, Skipped: 157
[INFO] Results:
[INFO]
[WARNING] Tests run: 273, Failures: 0, Errors: 0, Skipped: 41

Time taken: 47 mins 38 secs.

anmolanmol1234 commented on hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStreamUtils.java:
```diff
@@ -0,0 +1,34 @@
+
+public class AbfsOutputStreamUtils {
+
+  private AbfsOutputStreamUtils() {
```
This new class can be avoided as there it supports only a single function call

snvijaya replied:
+1 , its a direct call to a method of passed in parameter.

saxenapranav replied:
Have removed the class.

anmolanmol1234 commented on hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/ITestAzureBlobFileSystemAppend.java:
```diff
@@ -603,4 +609,94 @@ public void testAppendWithLease() throws Exception {
+   * The next close would also fail for the last caught exception.
+   */
+  @Test
+  public void testIntermittentAppendFailureToBeReported() throws Exception {
```
Can this test be broken into smaller tests based on the 3 scenarios listed in the description section ?

saxenapranav replied:
Taken.

snvijaya commented:
As discussed in the past, instead of a elaborate PR description which requires reviewer to do a code browsing in description, lets make sure that the correct code spots have the mention of when last error is saved and when it will be sent to called when a new API call is hit.

saxenapranav commented:
1. Removed the test-util class
2. separate the test method into two as per the separate logics.
3. added comments in src code on how failures are handled in the lifetime of outputstream.

saxenapranav commented:
Example of hflush returning the exception:
In append following exception was raised: 
```
throw new AbfsRestOperationException(503, "", "", new Exception("Mocked exception"));
```

exception caught in hflush:
```
java.io.IOException: Status code: 503 error code:  error message: java.lang.Exception: Mocked exception

	at org.apache.hadoop.fs.azurebfs.services.AbfsOutputStream.waitForAppendsToComplete(AbfsOutputStream.java:808)
	at org.apache.hadoop.fs.azurebfs.services.AbfsOutputStream.flushWrittenBytesToService(AbfsOutputStream.java:823)
	at org.apache.hadoop.fs.azurebfs.services.AbfsOutputStream.flushInternal(AbfsOutputStream.java:718)
	at org.apache.hadoop.fs.azurebfs.services.AbfsOutputStream.hflush(AbfsOutputStream.java:638)
	at org.apache.hadoop.fs.FSDataOutputStream.hflush(FSDataOutputStream.java:136)
	at org.apache.hadoop.fs.azurebfs.ITestAzureBlobFileSystemAppend.testWriteAsyncOpFailedAfterCloseCalled(ITestAzureBlobFileSystemAppend.java:732)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:750)
Caused by: Status code: 503 error code:  error message: java.lang.Exception: Mocked exception
	at org.apache.hadoop.fs.azurebfs.ITestAzureBlobFileSystemAppend.lambda$testWriteAsyncOpFailedAfterCloseCalled$17(ITestAzureBlobFileSystemAppend.java:698)
	at org.mockito.internal.stubbing.StubbedInvocationMatcher.answer(StubbedInvocationMatcher.java:42)
	at org.mockito.internal.handler.MockHandlerImpl.handle(MockHandlerImpl.java:103)
	at org.mockito.internal.handler.NullResultGuardian.handle(NullResultGuardian.java:29)
	at org.mockito.internal.handler.InvocationNotifierHandler.handle(InvocationNotifierHandler.java:34)
	at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor.doIntercept(MockMethodInterceptor.java:82)
	at org.mockito.internal.creation.bytebuddy.MockMethodAdvice.handle(MockMethodAdvice.java:134)
	at org.apache.hadoop.fs.azurebfs.services.AbfsClient.append(AbfsClient.java:850)
	at org.apache.hadoop.fs.azurebfs.services.AbfsOutputStream.lambda$uploadBlockAsync$0(AbfsOutputStream.java:471)
	at org.apache.hadoop.thirdparty.com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:125)
	at org.apache.hadoop.thirdparty.com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:69)
	at org.apache.hadoop.thirdparty.com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:78)
	at org.apache.hadoop.util.SemaphoredDelegatingExecutor$RunnableWithPermitRelease.run(SemaphoredDelegatingExecutor.java:196)
	at org.apache.hadoop.util.SemaphoredDelegatingExecutor$RunnableWithPermitRelease.run(SemaphoredDelegatingExecutor.java:196)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)

```

snvijaya reviewed (approved):
+1
